src/models/db/case.js

```diff
@@ -1,72 +1,163 @@
+const co = require('co')
+const assert = require('assert')
 
-      data.expires_at = new Date(data.expires_at)
+const TABLE_NAME = 'N_CASES'
+const STATUS_TABLE_NAME = 'N_LU_CASE_STATUS'
 
-      return data
-    }
+function getTransaction (options) {
+  return !options ? knex : (!options.transaction ? knex : options.transaction)
+}
 
-    static convertToExternal (data) {
-      data.id = uri.make('case', data.id.toLowerCase())
-      data.notaries = [config.getIn(['server', 'base_uri'])]
-      data.expires_at = moment(data.expires_at).toISOString() // format('YYYY-MM-DD HH:mm:ss');
-      delete data.Notaries
-      if (!data.exec_cond_fulfillment) {
-        delete data.exec_cond_fulfillment
-      }
-      return data
-    }
+function transaction (callback) {
+  return knex.transaction(co.wrap(callback))
+}
 
-    static convertFromPersistent (data) {
-      data.execution_condition = JSON.parse(data.execution_condition)
-      if (data.exec_cond_fulfillment) {
-        data.exec_cond_fulfillment = JSON.parse(data.exec_cond_fulfillment)
-      }
-      data.notification_targets = JSON.parse(data.notification_targets)
-      delete data.created_at
-      delete data.updated_at
-      if (!Array.isArray(data.notaries)) {
-        data.notaries = [ data.notaries ]
+function * insertCases (caseObjects, options) {
+  if (caseObjects.length <= 0) {
+    return
+  }
+  return Promise.all(_.map(caseObjects, (caseObject) => {
+    return convertToPersistent(caseObject, options)
+  }))
+  .then((persistentCases) => {
+    return getTransaction(options)(TABLE_NAME)
+      .insert(persistentCases).then()
+  })
+}
+
+function insertCase (caseObject, options) {
+  return convertToPersistent(caseObject, options)
+    .then((persistentCase) => {
+      return getTransaction(options)(TABLE_NAME).insert(persistentCase).then()
+    })
+}
+
+function updateCase (caseObject, options) {
+  return convertToPersistent(caseObject, options)
+    .then((persistentCase) => {
+      assert(persistentCase.CASE_ID, 'Cannot update a case without a primary key: ' +
+        JSON.stringify(persistentCase))
+      return getTransaction(options)(TABLE_NAME).update(persistentCase)
+      .where('CASE_ID', persistentCase.CASE_ID)
+      .then()
+    })
+}
+
+function * _upsertCase (caseData, options) {
+  assert(options.transaction)
+  const existingCase = yield getCaseByUuid(caseData.id, options)
+  if (existingCase) {
+    yield updateCase(_.merge({}, existingCase, caseData), options)
+  } else {
+    yield insertCase(caseData, options)
+  }
+  return Boolean(existingCase)
+}
+
+function * upsertCase (caseData, options) {
+  if (options && options.transaction) {
+    return yield _upsertCase(caseData, options)
+  } else {
+    let result
+    yield transaction(function * (transaction) {
+      result = yield _upsertCase(caseData,
+        _.assign({}, options || {}, {transaction}))
+    })
+    return result
+  }
+}
+
+function getByKey (key, value, options) {
```

Comment:
It might be worth using this helper module to separate out the lower-level database code: https://github.com/interledger/five-bells-ledger/blob/master/src/models/db/utils.js